### PR TITLE
Add status field to Interface

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes: #984 
# What's Changed
Add status field to Interface Model

# Checklist 

- [x] status is a new mandatory field on Interface
- [x] REST API would need to be versioned
- [x] Interface Statuses are: Active, Planned, Maintenance, Failed, and Decommissioned
- [x] Default value for migration, legacy API would be "Active"

### Interface List UI
Screenshot 2022-05-05 at 12 52 15 PM

### Interface Detail UI
Screenshot 2022-05-05 at 12 52 40 PM

### Device Interface List UI
Screenshot 2022-05-05 at 12 53 14 PM

# TODO

- [x] Explanation of Change(s)
- [x] Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/dcim/choices.py [hunk 1]
```diff
@@ -909,6 +909,31 @@ class InterfaceModeChoices(ChoiceSet):
...
+
+    CHOICES = (
+        (STATUS_PLANNED, "Planned"),
+        (STATUS_FAILED, "Failed"),
+        (STATUS_ACTIVE, "Active"),
+        (STATUS_DECOMMISSIONING, "Decommissioning"),
+        (STATUS_MAINTENANCE, "Maintenance"),
+    )
+
+    CSS_CLASSES = {
```
Comment (glennmatthews) [inline nautobot/dcim/choices.py:928]: Is CSS_CLASSES actually used/needed anywhere? If not I'd say let's not add it.

File: nautobot/dcim/models/device_components.py [hunk 2]
```diff
@@ -598,6 +601,7 @@ def to_csv(self):
             self.mgmt_only,
             self.description,
             self.get_mode_display(),
+            self.get_status_display(),
         )
 
     def clean(self):
```
Comment (glennmatthews) [inline nautobot/dcim/models/device_components.py:607]: Thinking about this from a not-breaking-existing-usage standpoint, would it make sense for `clean()` and/or `save()` to automatically set `self.status = status.objects.get(slug="active")` (or equivalent) if an Interface or VMInterface is instantiated without setting its status? Thinking about the impact to Jobs in particular here.

File: nautobot/virtualization/migrations/0006_vminterface_status_data_migration.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
+# Generated by Django 3.1.14 on 2022-03-19 23:21
```
Comment (glennmatthews) [inline nautobot/virtualization/migrations/0006_vminterface_status_data_migration.py:1]: This migration should be numbered 0007 shouldn't it?

File: nautobot/dcim/choices.py (hunk 1 above)
Comment (timizuoebideri1) [reply nautobot/dcim/choices.py:928]: Its been used by [export_statuses_from_choiceset()](https://github.com/nautobot/nautobot/blob/timizuo-GH984-interface-status/nautobot/extras/management/__init__.py#L80), [2bbef5af517R15](https://github.com/nautobot/nautobot/pull/1753/files#diff-0d98c24df83f4a421cf3962f7a1bc886836279d0b9879439d4a8b2bbef5af517R15)

File: nautobot/dcim/api/serializers.py [hunk 4]
```diff
@@ -1083,6 +1085,40 @@ def validate(self, data):
         return super().validate(data)
 
 
+class InterfaceSerializerVersion13(InterfaceSerializer, StatusModelSerializerMixin):
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:1088]: Personally I'd prefer if `InterfaceSerializer` stays the current/latest version of the serializer, and we add `InterfaceSerializerVersion12` as the frozen legacy serializer version.

File: nautobot/dcim/migrations/0011_interface_status_data_migration.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    ContentType = app.get_model("contenttypes.ContentType")
+
+    interface_content_type = ContentType.objects.get_for_model(Interface)
+
+    for interface in Interface.objects.filter(status__slug=InterfaceStatusChoices.STATUS_ACTIVE):
+        interface.status = None
+        interface.save()
+
+    for status in Status.objects.filter(content_types__in=[interface_content_type]):
+        status.content_types.clear()
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0011_interface_status_data_migration.py:50]: This seems wrong - won't it clear *all* content_types from the affected statuses, rather than just removing the interface_content_type?

File: nautobot/dcim/templates/dcim/interface_edit.html [hunk 6]
```diff
@@ -17,6 +17,7 @@
             {% endif %}
             {% render_field form.name %}
             {% render_field form.label %}
+            {% render_field form.status %}
```
Comment (glennmatthews) [inline nautobot/dcim/templates/dcim/interface_edit.html:20]: My preference, at least:

```suggestion
            {% render_field form.status %}
            {% render_field form.label %}
```

File: nautobot/dcim/models/device_components.py [hunk 7]
```diff
@@ -501,6 +504,10 @@ class Meta:
 
     def save(self, *args, **kwargs):
 
+        # set interface status to active if status not provided
+        if not self.status:
+            self.status = Status.objects.get(slug=InterfaceStatusChoices.STATUS_ACTIVE)
```
Comment (glennmatthews) [inline nautobot/dcim/models/device_components.py:509]: Maybe something like this would be safer?

```suggestion
        if not self.status:
            interface_content_type = ContentType.objects.get_for_model(Interface)
            try:
                self.status = Status.objects.get(slug=InterfaceStatusChoices.STATUS_ACTIVE, content_types__in=[interface_content_type])
            except Status.DoesNotExist:
                self.status = Status.objects.filter(content_types__in=[interface_content_type]).first()
```

File: nautobot/dcim/tests/test_api.py [hunk 8]
```diff
@@ -1490,6 +1490,58 @@ def setUpTestData(cls):
...
+    def setUpTestData(cls):
+        manufacturer = Manufacturer.objects.create(name="Test Manufacturer 1", slug="test-manufacturer-1")
+        devicetype = DeviceType.objects.create(manufacturer=manufacturer, model="Device Type 1", slug="device-type-1")
+        site = Site.objects.create(name="Site 1", slug="site-1")
+        devicerole = DeviceRole.objects.create(name="Test Device Role 1", slug="test-device-role-1", color="ff0000")
+        device = Device.objects.create(device_type=devicetype, device_role=devicerole, name="Device 1", site=site)
+
+        Interface.objects.create(device=device, name="Interface 1", type="1000base-t")
+        Interface.objects.create(device=device, name="Interface 2", type="1000base-t")
+        Interface.objects.create(device=device, name="Interface 3", type="1000base-t")
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_api.py:1506]: Would probably make sense to explicitly set `status` on these created Interfaces in this test case at least.

File: nautobot/virtualization/api/serializers.py [hunk 9]
```diff
@@ -252,3 +254,34 @@ def validate(self, data):
                 )
 
         return super().validate(data)
+
+
+class VMInterfaceSerializerVersion13(TaggedObjectSerializer, ValidatedModelSerializer, StatusModelSerializerMixin):
```
Comment (glennmatthews) [inline nautobot/virtualization/api/serializers.py:259]: Same suggestion here about `VMInterfaceSerializer` and `VMInterfaceSerializerVersion12`.

File: nautobot/virtualization/api/views.py [hunk 10]
```diff
@@ -93,8 +94,34 @@ def get_serializer_class(self):
...
+    def get_serializer_class(self):
+        if (
+            not self.brief
+            and not getattr(self, "swagger_fake_view", False)
+            and (
+                not hasattr(self.request, "major_version")
+                or self.request.major_version > 1
+                or (self.request.major_version == 1 and self.request.minor_version < 3)
+            )
+        ):
```
Comment (glennmatthews) [inline nautobot/virtualization/api/views.py:123]: We're probably reaching a point where it would be desirable to add a helper method on `ModelViewSet` that could encapsulate this kind of repeated logic.

File: nautobot/virtualization/forms.py [hunk 11]
```diff
@@ -48,6 +48,7 @@
 #
 # Cluster types
 #
+from ..extras.models import Status
```
Comment (glennmatthews) [inline nautobot/virtualization/forms.py:51]:
```suggestion
from nautobot.extras.models import Status
```

File: nautobot/virtualization/templates/virtualization/vminterface.html [hunk 12]
```diff
@@ -17,6 +17,12 @@
                     <td>Name</td>
                     <td>{{ object.name }}</td>
                 </tr>
+                <tr>
+                        <td>Status</td>
+                        <td>
+                            <span class="label" style="color: {{ object.status.color|fgcolor }}; background-color: #{{ object.status.color }}">{{ object.get_status_display }}</span>
```
Comment (glennmatthews) [inline nautobot/virtualization/templates/virtualization/vminterface.html:23]: I'm a bit surprised we don't yet have a standard templatetag for rendering Status objects...

File: nautobot/virtualization/tests/test_api.py [hunk 13]
```diff
@@ -326,3 +326,57 @@ def setUpTestData(cls):
...
+    @classmethod
+    def setUpTestData(cls):
+
+        clustertype = ClusterType.objects.create(name="Test Cluster Type 1", slug="test-cluster-type-1")
+        cluster = Cluster.objects.create(name="Test Cluster 1", type=clustertype)
+        virtualmachine = VirtualMachine.objects.create(cluster=cluster, name="Test VM 1")
+
+        VMInterface.objects.create(virtual_machine=virtualmachine, name="Interface 1")
+        VMInterface.objects.create(virtual_machine=virtualmachine, name="Interface 2")
+        VMInterface.objects.create(virtual_machine=virtualmachine, name="Interface 3")
```
Comment (glennmatthews) [inline nautobot/virtualization/tests/test_api.py:349]: Would be good to set `status` on these objects.

Comment (timizuoebideri1) [reply nautobot/virtualization/tests/test_api.py:349]: `status` is set to `active` if none was provided.

Comment (glennmatthews) [reply nautobot/virtualization/tests/test_api.py:349]: Understood, but it would be good to set explicitly in the "new" test cases since the default behavior is really only provided for backwards-compatibility.

File: nautobot/dcim/api/serializers.py [hunk 14]
```diff
@@ -1015,7 +1015,7 @@ class Meta:
         opt_in_fields = ["computed_fields"]
 
 
-class InterfaceSerializer(
+class InterfaceSerializerVersion12(
```
Comment (bryanculver) [inline nautobot/dcim/api/serializers.py:1018]: Let's add a `TODO` here to collapse this with `InterfaceSerializer` in 2.0.

Unfortunately because of the mixin needs for the newer version we can't do something like https://github.com/nautobot/nautobot/blob/develop/nautobot/ipam/api/serializers.py#L443-L444

File: nautobot/utilities/utils.py [hunk 15]
```diff
@@ -530,3 +530,26 @@ def get_table_for_model(model):
 
 # Setup UtilizationData named tuple for use by multiple methods
 UtilizationData = namedtuple("UtilizationData", ["numerator", "denominator"])
+
+
+def versioned_serializer_selector(obj, legacy_serializer, serializer):
```
Comment (bryanculver) [inline nautobot/utilities/utils.py:535]: Love this cleanup but the version detection needs to be flexible. Maybe this can take a dictionary for the "introduced in" version, or at least as an argument. Because for the new serializer for Interface status, that needs to be in 1.4.

**Review by bryanculver — CHANGES_REQUESTED** [review]

File: nautobot/utilities/utils.py (hunk 15 above)
Comment (bryanculver) [reply nautobot/utilities/utils.py:535]: I'm thinking a dictionary like this:
```py
serializer_choices={
  "1.3": legacy_serializer=serializers.InterfaceSerializerVersion12,
  "1.4": serializer=super().get_serializer_class(), 
}
```

Where the key is the "version introduced in". (one footnote is that our current InterfaceSerializer today should be Version12, but the new one being implemented here is 1.4 since that's the version it's being introduced in)

We might need to workshop this a bit to help determine what the default is, etc. I love the code reduction but it's not going to scale as we progress with more changes.

File: nautobot/virtualization/tests/test_api.py [hunk 16]
```diff
@@ -326,3 +327,54 @@ def setUpTestData(cls):
                 "untagged_vlan": vlans[2].pk,
             },
         ]
+
+
+class VMInterfaceTestVersion13(VMInterfaceTestVersion12):
+    api_version = "1.3"
```
Comment (bryanculver) [inline nautobot/virtualization/tests/test_api.py:333]:
```suggestion
class VMInterfaceTestVersion14(VMInterfaceTestVersion12):
    api_version = "1.4"
```

This will begin to be available in 1.4. More places need to be updated

File: nautobot/virtualization/templates/virtualization/vminterface.html (hunk 12 above)
Comment (timizuoebideri1) [reply nautobot/virtualization/templates/virtualization/vminterface.html:23]: I guess creating a template tag for rending the `Status` object is out of scope for this issue right?

Comment (glennmatthews) [reply nautobot/virtualization/templates/virtualization/vminterface.html:23]: Yeah - but it would be good to open a housekeeping issue.

File: nautobot/dcim/api/serializers.py [hunk 17]
```diff
@@ -1015,7 +1017,8 @@ class Meta:
         opt_in_fields = ["computed_fields"]
 
 
-class InterfaceSerializer(
+# TODO: collapse this with InterfaceSerializer in 2.0.
+class InterfaceSerializerVersion13(
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:1021]: I think this would better be named as `Version12` as it's in use by both the v1.2 and v1.3 REST API?

File: nautobot/dcim/api/serializers.py [hunk 18]
```diff
@@ -1069,6 +1072,16 @@ class Meta:
...
 
+        # set interface status to active if status not provided
+        if not data.get("status"):
+            # status is currently required in the Interface model but not required in api_version < 1.3 serializers
+            # which raises an error when validating except status is explicitly set here
+            query = Status.objects.get_for_model(Interface)
+            try:
+                data["status"] = query.get(slug=InterfaceStatusChoices.STATUS_ACTIVE)
+            except Status.DoesNotExist:
+                data["status"] = query.first()
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:1083]: To @jathanism's comment earlier - in the case where Active status doesn't exist, better to have a fallback to the first (arbitrary) status and log a warning? Or better to report an error and require that the user either recreate the Active status or switch to the newer API version that lets them explicitly choose a status? I know I suggested the fallback originally but now I'm reconsidering - sorry!

File: nautobot/dcim/api/views.py [hunk 19]
```diff
@@ -593,14 +593,37 @@ class PowerOutletViewSet(PathEndpointMixin, CustomFieldModelViewSet):
...
 
+    def get_serializer_class(self):
+        serializer_choices = (
+            SerializerVersions(versions=["1.2", "1.3"], serializer=serializers.InterfaceSerializerVersion13),
+        )
+        return versioned_serializer_selector(
+            obj=self,
+            serializer_choices=serializer_choices,
+            current_serializer=super().get_serializer_class(),
+        )
```
Comment (glennmatthews) [inline nautobot/dcim/api/views.py:625]: Maybe just me, but this API seems rather unintuitive.

- Why does `SerializerVersions` (plural) take a single serializer as an input? Would calling it something like "SerializerForAPIVersions(versions, serializer)` be clearer?
- How does `current_serializer` relate to the above versions? Is this the serializer that gets used if none of the serializer_choices match the current API version, or something else?

I personally would have expected something like the below pseudocode instead:

```
serializer_choices = {
    "1.2": serializers.InterfaceSerializerVersion12,
    "1.3": serializers.InterfaceSerializerVersion12,
    "latest": serializers.InterfaceSerializer
}
return versioned_serializer_selector(serializer_choices, self)
```

File: nautobot/dcim/migrations/0011_interface_status_data_migration.py [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+    interface_content_type = ContentType.objects.get_for_model(Interface)
+    choices = export_statuses_from_choiceset(InterfaceStatusChoices)
+
+    # Create Interface Statuses and add dcim.Interface to its content_types
+    for choice_kwargs in choices:
+        try:
+            obj, created = Status.objects.get_or_create(**choice_kwargs)
+        except IntegrityError:
+            choice_kwargs.pop("color")
+            obj, created = Status.objects.get_or_create(**choice_kwargs)
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0011_interface_status_data_migration.py:23]: Why is this needed? Perhaps add an explanatory comment?

File: nautobot/dcim/migrations/0011_interface_status_data_migration.py [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+        except Exception as err:
+            raise SystemExit(
+                f"Unexpected error while running data migration to populate" f"status for dcim.interface: {err}"
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0011_interface_status_data_migration.py:26]:
```suggestion
                f"Unexpected error while running data migration to populate status for dcim.interface: {err}"
```

File: nautobot/dcim/models/device_components.py [hunk 22]
```diff
@@ -501,6 +504,15 @@ class Meta:
 
     def save(self, *args, **kwargs):
 
+        # set interface status to active if status not provided
+        if not self.status:
+            query = Status.objects.get_for_model(self)
+            try:
+                status = query.get(slug=InterfaceStatusChoices.STATUS_ACTIVE)
+            except Status.DoesNotExist:
+                status = query.first()
```
Comment (glennmatthews) [inline nautobot/dcim/models/device_components.py:513]: Same thought here - changing my mind from earlier, I now think it's probably better if we bubble up an appropriate exception if STATUS_ACTIVE doesn't exist, rather than picking an arbitrary/random status to use.

File: nautobot/dcim/tests/test_api.py [hunk 23]
```diff
@@ -1435,14 +1436,14 @@ def setUpTestData(cls):
         ]
 
 
-class InterfaceTest(Mixins.ComponentTraceMixin, APIViewTestCases.APIViewTestCase):
+class InterfaceTestVersion13(Mixins.ComponentTraceMixin, APIViewTestCases.APIViewTestCase):
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_api.py:1439]: As in my earlier comment, I think maybe Version12 would be more accurate here.

File: nautobot/dcim/tests/test_forms.py [hunk 24]
```diff
@@ -159,11 +159,13 @@ def setUpTestData(cls):
 
     def test_interface_label_count_valid(self):
         """Test that a `label` can be generated for each generated `name` from `name_pattern` on InterfaceCreateForm"""
+        status_active = Status.objects.get_for_model(Interface).get(slug=InterfaceStatusChoices.STATUS_ACTIVE)
         interface_data = {
             "device": self.device.pk,
             "name_pattern": "eth[0-9]",
             "label_pattern": "Interface[0-9]",
             "type": InterfaceTypeChoices.TYPE_100ME_FIXED,
+            "status": status_active,
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_forms.py:168]: Should this instead be?

```suggestion
            "status": status_active.pk,
```

File: nautobot/docs/models/dcim/interface.md [hunk 25]
```diff
@@ -1,6 +1,12 @@
 # Interfaces
 
-Interfaces in Nautobot represent network interfaces used to exchange data with connected devices. On modern networks, these are most commonly Ethernet, but other types are supported as well. Each interface must be assigned a type, and may optionally be assigned a MAC address, MTU, and IEEE 802.1Q mode (tagged or access). Each interface can also be enabled or disabled, and optionally designated as management-only (for out-of-band management).
+Interfaces in Nautobot represent network interfaces used to exchange data with connected devices. On modern networks, these are most commonly Ethernet, but other types are supported as well. Each interface must be assigned a type, an operational [`status`](https://nautobot.readthedocs.io/en/stable/models/extras/status/) and may optionally be assigned a MAC address, MTU, and IEEE 802.1Q mode (tagged or access). Each interface can also be enabled or disabled, and optionally designated as management-only (for out-of-band management). The following operational statuses are available by default:
+
```
Comment (glennmatthews) [inline nautobot/docs/models/dcim/interface.md:4]:
```suggestion
Interfaces in Nautobot represent network interfaces used to exchange data with connected devices. On modern networks, these are most commonly Ethernet, but other types are supported as well. Each interface must be assigned a type, an operational [`status`](https://nautobot.readthedocs.io/en/stable/models/extras/status/) and may optionally be assigned a MAC address, MTU, and IEEE 802.1Q mode (tagged or access). Each interface can also be enabled or disabled, and optionally designated as management-only (for out-of-band management).

_The `status` field was added in version 1.4.0_

The following operational statuses are available by default:

```

File: nautobot/docs/models/virtualization/vminterface.md [hunk 26]
```diff
@@ -1,3 +1,11 @@
 # Interfaces
 
-Virtual machine interfaces behave similarly to device interfaces, and can be assigned IP addresses, VLANs, and services. However, given their virtual nature, they lack properties pertaining to physical attributes. For example, VM interfaces do not have a physical type and cannot have cables attached to them.
+Virtual machine interfaces behave similarly to device interfaces, and can be assigned IP addresses, VLANs, an operational [`status`](https://nautobot.readthedocs.io/en/stable/models/extras/status/) and services. However, given their virtual nature, they lack properties pertaining to physical attributes. For example, VM interfaces do not have a physical type and cannot have cables attached to them.
+
+The following operational statuses are available by default:
```
Comment (glennmatthews) [inline nautobot/docs/models/virtualization/vminterface.md:5]:
```suggestion
_The `status` field was added in version 1.4.0_

The following operational statuses are available by default:
```

File: nautobot/virtualization/api/serializers.py [hunk 27]
```diff
@@ -240,6 +243,16 @@ class Meta:
 
     def validate(self, data):
 
+        # set vminterface status to active if status not provided
+        if not data.get("status"):
+            # status is currently required in the VMInterface model but not required in api_version < 1.3 serializers
```
Comment (glennmatthews) [inline nautobot/virtualization/api/serializers.py:248]:
```suggestion
            # status is currently required in the VMInterface model but not required in api_version < 1.4 serializers
```

File: nautobot/dcim/api/serializers.py [hunk 28]
```diff
@@ -1069,6 +1072,16 @@ class Meta:
 
     def validate(self, data):
 
+        # set interface status to active if status not provided
+        if not data.get("status"):
+            # status is currently required in the Interface model but not required in api_version < 1.3 serializers
+            # which raises an error when validating except status is explicitly set here
+            query = Status.objects.get_for_model(Interface)
+            try:
+                data["status"] = query.get(slug=InterfaceStatusChoices.STATUS_ACTIVE)
```
Comment (glennmatthews) [inline nautobot/dcim/api/serializers.py:1081]: To ensure correctness, we should probably also filter on `content_types` - if someone edited the Active status  so that it's no longer applicable to Interface records, we don't want to create invalid data here. Same comment on other places where you have this logic.

Comment (timizuoebideri1) [reply nautobot/dcim/api/serializers.py:1081]: Not needed since `Status.objects.get_for_model(Interface)` takes care of that

File: nautobot/dcim/migrations/0011_interface_status_data_migration.py (hunk 20 above)
Comment (timizuoebideri1) [reply nautobot/dcim/migrations/0011_interface_status_data_migration.py]: Thinking of it, I don't think is needed

File: nautobot/dcim/api/views.py (hunk 19 above)
Comment (timizuoebideri1) [reply nautobot/dcim/api/views.py]: > ```
> serializer_choices = {
>     "1.2": serializers.InterfaceSerializerVersion12,
>     "1.3": serializers.InterfaceSerializerVersion12,
>     "latest": serializers.InterfaceSerializer
> }
> return versioned_serializer_selector(serializer_choices, self)
> ```

I don't think this approach is great, just imagine having several APIs using the same serializer, too much repetition IMO
```
 serializer_choices = {
    "1.2": serializers.InterfaceSerializerVersion12,
    "1.3": serializers.InterfaceSerializerVersion12,
    "1.4": serializers.InterfaceSerializerVersion12,
    "1.5": serializers.InterfaceSerializerVersion12,
    "1.6": serializers.InterfaceSerializerVersion12,
    "1.7": serializers.InterfaceSerializerVersion12,
    "latest": serializers.InterfaceSerializer
 }
```

File: nautobot/utilities/utils.py [hunk 29]
```diff
@@ -530,3 +530,38 @@ def get_table_for_model(model):
 
 # Setup UtilizationData named tuple for use by multiple methods
 UtilizationData = namedtuple("UtilizationData", ["numerator", "denominator"])
+
+# namedtuple accepts versions(list of API versions) and serializer(Appropriate Serializer class for versions).
```
Comment (timizuoebideri1) [inline nautobot/utilities/utils.py:534]:
```suggestion
# namedtuple accepts versions(list of API versions) and serializer(Related 'versions' serializer).
```

**Review by glennmatthews — APPROVED** [review]

File: nautobot/docs/models/dcim/interface.md [hunk 30]
```diff
@@ -1,6 +1,16 @@
 # Interfaces
 
-Interfaces in Nautobot represent network interfaces used to exchange data with connected devices. On modern networks, these are most commonly Ethernet, but other types are supported as well. Each interface must be assigned a type, and may optionally be assigned a MAC address, MTU, and IEEE 802.1Q mode (tagged or access). Each interface can also be enabled or disabled, and optionally designated as management-only (for out-of-band management).
+Interfaces in Nautobot represent network interfaces used to exchange data with connected devices. On modern networks, these are most commonly Ethernet, but other types are supported as well. Each interface must be assigned a type, an operational [`status`](https://nautobot.readthedocs.io/en/stable/models/extras/status/) and may optionally be assigned a MAC address, MTU, and IEEE 802.1Q mode (tagged or access). Each interface can also be enabled or disabled, and optionally designated as management-only (for out-of-band management).
+
+_The `status` field was added in version 1.4.0_
```
Comment (bryanculver) [inline nautobot/docs/models/dcim/interface.md:5]:
```suggestion
_Added in 1.4.0: `status` field_
```

Comment (bryanculver) [inline nautobot/docs/models/dcim/interface.md:5]:
```suggestion
_Added in 1.4.0: `status` field_
```

File: nautobot/docs/models/virtualization/vminterface.md [hunk 31]
```diff
@@ -1,3 +1,13 @@
 # Interfaces
 
-Virtual machine interfaces behave similarly to device interfaces, and can be assigned IP addresses, VLANs, and services. However, given their virtual nature, they lack properties pertaining to physical attributes. For example, VM interfaces do not have a physical type and cannot have cables attached to them.
+Virtual machine interfaces behave similarly to device interfaces, and can be assigned IP addresses, VLANs, an operational [`status`](https://nautobot.readthedocs.io/en/stable/models/extras/status/) and services. However, given their virtual nature, they lack properties pertaining to physical attributes. For example, VM interfaces do not have a physical type and cannot have cables attached to them.
+
+_The `status` field was added in version 1.4.0_
```
Comment (bryanculver) [inline nautobot/docs/models/virtualization/vminterface.md:5]:
```suggestion
_Added in 1.4.0: `status` field_
```
